sdv/single_table/utils.py

```diff
@@ -27,24 +31,37 @@ def detect_discrete_columns(metadata, data):
             A list of discrete columns to be used with some of ``sdv`` synthesizers.
     """
     discrete_columns = []
-
     for column in data.columns:
-        if column in metadata.columns:
-            if metadata.columns[column]['sdtype'] not in ['numerical', 'datetime']:
-                discrete_columns.append(column)
+        # Numerical and datetime columns never get preprocessed into categorical ones
+        if column in metadata.columns and \
+                metadata.columns[column]['sdtype'] in ['numerical', 'datetime']:
+            continue
 
-        else:
-            column_data = data[column].dropna()
-            if set(column_data.unique()) == {0.0, 1.0}:
-                column_data = column_data.astype(bool)
+        column_data = data[column].dropna()
 
-            try:
-                dtype = column_data.infer_objects().dtype.kind
-                if dtype in ['O', 'b']:
-                    discrete_columns.append(column)
+        # Ignore columns with only nans and empty datasets
+        if column_data.empty:
+            continue
 
-            except Exception:
-                discrete_columns.append(column)
+        # Non-integer floats and integers with too many unique values cannot be categorical
+        try:
+            column_data = column_data.astype('float')
+            rounded_column_data = list(column_data.round())
+            column_data = list(column_data)
+            is_float = column_data != rounded_column_data
+            is_int = column_data == rounded_column_data
```

Comment:
I don't think we need to compute `is_float` and `is_int` since one is the opposite of the other. Also is it more efficient to do these comparisons as lists than dataframes/series? I would've that it would be faster the other way